### PR TITLE
Add FileHeaderRule

### LINKED ISSUE #844: File header rule
We could have an opt-in rule that warned when the file have the default comment header included in Xcode template. For example:

```
//
//  LegacyConstructorRule.swift
//  SwiftLint
//
//  Created by Marcelo Fabri on 29/11/15.
//  Copyright © 2015 Realm. All rights reserved.
//
```

This wouldn't be enabled for SwiftLint 😅

### PR DESCRIPTION
**Pull request by marcelofabri**
Fixes #844

### CODE REVIEW

**Comment by jpsim** [comment]
Is there a way we can make this rule a bit more generic? Maybe by making it configurable by things that must or must not be in the top comment?

For example, as much as some people might consider the top comment to be superfluous, some projects _require_ an identical header (e.g. [Realm](https://github.com/realm/realm-cocoa/blob/master/RealmSwift/Realm.swift#L1-L17)). And some use a consistent template (SwiftLint).

**Comment by marcelofabri** [comment]
Maybe we could add two configurations, like `enforced_pattern` and `forbidden_pattern`?

**Comment by jpsim** [comment]
Maybe `required_string`, `required_pattern`, `forbidden_string` and `forbidden_pattern`?

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,6 +29,11 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#910](https://github.com/realm/SwiftLint/issues/910)
   [#911](https://github.com/realm/SwiftLint/issues/911)
+  
+* Add `file_header` opt-in rule that warns when a file contains a
```
Comment (jpsim) [inline CHANGELOG.md]: This is now in the wrong changelog section.

File: Source/SwiftLintFramework/Models/MasterRuleList.swift [hunk 2]
```diff
@@ -63,6 +63,7 @@ public let masterRuleList = RuleList(rules:
     LegacyConstantRule.self,
     LegacyConstructorRule.self,
     LegacyNSGeometryFunctionsRule.self,
+    FileHeaderRule.self,
```
Comment (jpsim) [inline Source/SwiftLintFramework/Models/MasterRuleList.swift]: This isn't alphabetized.

File: Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            return nil
+        }
+        set {
+            _forbiddenRegex = newValue
+        }
+    }
+
+    // swiftlint:disable:next force_try
+    private static let defaultRegex = try! NSRegularExpression(pattern: "\\bCopyright\\b",
+                                                              options: [.caseInsensitive])
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift]: `options` should be vertically aligned with `pattern`.

File: Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    public var consoleDescription: String {
+        return severityConfiguration.consoleDescription + ", required_string: \(requiredString)" +
+        ", required_pattern: \(requiredPattern), forbidden_string: \(forbiddenString)" +
+        ", forbidden_pattern: \(forbiddenPattern)"
+    }
+
+    public init() {
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift]: `public init() {}`?

File: Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            throw ConfigurationError.unknownConfiguration
+        }
+
+        if let requiredString = configuration["required_string"] as? String {
+            self.requiredString = requiredString
+            requiredRegex = try NSRegularExpression(pattern: requiredString,
+                                                    options: [.ignoreMetacharacters])
+        } else if let requiredPattern = configuration["required_pattern"] as? String {
+            self.requiredPattern = requiredPattern
+            requiredRegex = try NSRegularExpression.cached(pattern: requiredPattern)
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift]: `requiredRegex = try .cached(pattern: requiredPattern)`

File: Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        if let forbiddenString = configuration["forbidden_string"] as? String {
+            self.forbiddenString = forbiddenString
+            forbiddenRegex = try NSRegularExpression(pattern: forbiddenString,
+                                                     options: [.ignoreMetacharacters])
+        } else if let forbiddenPattern = configuration["forbidden_pattern"] as? String {
+            self.forbiddenPattern = forbiddenPattern
+            forbiddenRegex = try NSRegularExpression.cached(pattern: forbiddenPattern)
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift]: `forbiddenRegex = try .cached(pattern: forbiddenPattern)`

**Comment by marcelofabri** [comment]
@jpsim I was enabling this rule for the project, but I've found (at least) two patterns:

```
//
//  String+SwiftLint.swift
//  SwiftLint
//
//  Created by JP Simard on 2015-05-16.
//  Copyright (c) 2015 Realm. All rights reserved.
//
```

```
//
//  SyntaxMap+SwiftLint.swift
//  SwiftLint
//
//  Created by Norio Nomura on 2/19/16.
//  Copyright © 2016 Realm. All rights reserved.
//
```

Should me make them consistent? The second one seems the be the Xcode 8.1 default template now. Or should I just make the regex more flexible and allow the first one as well?

**Comment by jpsim** [comment]
I'd love if you could address the inconsistencies, actually! Sticking with the newer Xcode 8.1 default seems preferable.

**Comment by marcelofabri** [comment]
@jpsim I've addressed the inconsistencies although the date in the template seems to change according to the user's locale. I've changed to what seems the default to the US locale.

File: Source/SwiftLintFramework/Rules/FileLengthRule.swift [hunk 7]
```diff
@@ -2,8 +2,8 @@
 //  FileLengthRule.swift
 //  SwiftLint
 //
-//  Created by JP Simard on 2015-05-16.
-//  Copyright (c) 2015 Realm. All rights reserved.
+//  Created by JP Simard on 5/16/15.
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/FileLengthRule.swift]: Some of these dates are now `mm/dd/yy` (like this one), others are now `dd/mm/yy`.

Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/FileLengthRule.swift]: they should be all in `mm/dd/yy` now

**Comment by jpsim** [comment]
👍 feel free to merge once CI agrees.

**Comment by marcelofabri** [comment]
@jpsim Do you have any idea on why the tests are passing with Xcode but not SPM? I get the same thing on my machine, so it's not a Travis issue.

**Comment by jpsim** [comment]
Looks like it's the `FunctionBodyLengthRuleTests` that are failing due to an unrelated rule (`file_header`). A quick glance seems to confirm my suspicions that these tests don't limit the rules it checks to the `FunctionBodyLengthRule`: https://github.com/realm/SwiftLint/blob/master/Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift

If you pass in a config constructed with [`makeConfig(...)`](https://github.com/realm/SwiftLint/blob/master/Tests/SwiftLintFrameworkTests/TestHelpers.swift#L116-L123) to `violations(...)` it should limit those tests to just the rule being tested.

**Comment by marcelofabri** [comment]
@jpsim I've created a config just disabling `file_header` as `makeConfig` was `fileprivate` and I'm not sure if we want to expose it. What do you think?

**Comment by jpsim** [comment]
I'd rather just make `makeConfig` `internal`.

**Comment by marcelofabri** [comment]
@jpsim do you wanna review the latest commits or can I merge this? 🚢

**Comment by marcelofabri** [comment]
🚢🎉

File: CHANGELOG.md [hunk 8]
```diff
@@ -14,6 +14,11 @@
   used as thousand separators in large numbers.  
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#924](https://github.com/realm/SwiftLint/issues/924)
+  
+* Add `file_header` opt-in rule that warns when a file contains a 
+  copyright comment header, such as the one Xcode adds by default.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#884](https://github.com/realm/SwiftLint/issues/884)
```
Comment (marcelofabri) [inline CHANGELOG.md]: @vincentisambart Thanks for letting me know! Fixed in #993

**Pull request merged**